src/network/eventloop.rs

```diff
@@ -1,21 +1,39 @@
-use crate::network::{
-    client::{Command, FileRequest, FileResponse},
-    swarm::{ComposedBehaviour, ComposedEvent},
+use crate::{
+    network::{
+        client::{Command, FileRequest, FileResponse},
+        swarm::{ComposedBehaviour, ComposedEvent},
+    },
+    workflow::receipt::Receipt,
 };
 use anyhow::{anyhow, Result};
+use either::Either;
 use libp2p::{
-    core::{either::EitherError, PeerId},
+    core::PeerId,
+    floodsub::{protocol::CodecError, FloodsubEvent},
     futures::StreamExt,
+    gossipsub::{error::GossipsubHandlerError, GossipsubEvent},
     kad::{GetProvidersOk, KademliaEvent, QueryId, QueryResult},
+    mdns,
     multiaddr::Protocol,
-    request_response::{RequestId, RequestResponseEvent, RequestResponseMessage, ResponseChannel},
+    request_response::{self, RequestId, ResponseChannel},
     swarm::{ConnectionHandlerUpgrErr, Swarm, SwarmEvent},
 };
 use std::{
     collections::{hash_map, HashMap, HashSet},
     io,
 };
 use tokio::sync::{mpsc, oneshot};
+use void::Void;
+
+type HandlerErr = Either<
+    Either<
+        Either<Either<GossipsubHandlerError, ConnectionHandlerUpgrErr<CodecError>>, io::Error>,
+        Void,
+    >,
+    ConnectionHandlerUpgrErr<io::Error>,
+>;
```

Comment:
Why the deeply nested `Either` instead of a custom error type and a way to lift errors into the sum?

Reply:
This is due to the underlying swarmevent errors (and what I was using from examples). The easier approach is a generic, which, I'll push.